File: cmsi-386/homework2/exercises.py
from random import sample
from math import pi
from cryptography.fernet import Fernet
from itertools import zip_longest
from requests import get


def change(cents):
    US_COIN_DENOMINATIONS = (25, 10, 5, 1)

    if cents < 0:
        raise ValueError('amount cannot be negative')

    result = []
    for d in US_COIN_DENOMINATIONS:
        result.append(cents // d)
        cents %= d

    return tuple(result)


def stretched(s):
    s = ''.join(s.split())
    return ''.join((c * (i+1)) for i, c in enumerate(s))


def scramble(s):
    return ''.join(sample(s, len(s)))


class Cylinder:
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height

    @property
    def volume(self):
        return pi * (self.radius ** 2) * self.height

    @property
    def surface_area(self):
        return 2 * pi * self.radius * (self.height + self.radius)

    def widen(self, factor):
        self.radius *= factor
        return self

    def stretch(self, factor):
        self.height *= factor
        return self


def powers(base, limit):
    power = 1
    while power <= limit:
        yield power
        power *= base
    return


def say(first=None):
    if first is None:
        return ''

    def sayNext(next=None):
        if next is None:
            return first
        return say(f'{first} {next}')

    return sayNext


def interleave(a, *b):
    return [item for pair in zip_longest(a, b) for item in pair if item is not None]


def make_crypto_functions(key):
    return ((cipher := Fernet(key)).encrypt, cipher.decrypt)


def top_ten_scorers(stats):
    flattened_stats = [[team, name, num_games, points] for team in stats.keys()
                       for name, num_games, points in stats[team]
                       if num_games >= 15]

    stats = [{
        'name': name,
        'ppg': points / num_games,
        'team': team
    } for team, name, num_games, points in flattened_stats]

    return sorted(stats, key=lambda player: player['ppg'], reverse=True)[:10]


def studio_ghibli_characters(*, hair_color, gender):
    query = {
        'hair_color': hair_color,
        'gender': gender
    }

    response = get('https://ghibliapi.herokuapp.com/people', params=query)

    response.raise_for_status()

    return [{k: v for k, v in person.items()
             if k in {'name', 'gender', 'age', 'eye_color', 'hair_color'}}
            for person in response.json()]

File: cmsi-386/homework1/src/exercises.js
import { createCipheriv, createDecipheriv } from 'crypto'
import fetch from 'node-fetch'

export function change(amount) {
  if (amount < 0) {
    throw new RangeError('amount cannot be negative.')
  }

  const US_COIN_DENOMINATIONS = [25, 10, 5, 1]
  const result = []

  US_COIN_DENOMINATIONS.forEach(d => {
    result.push(Math.floor(amount / d))
    amount %= d
  })

  return result
}

export function stretched(str) {
  let result = ''

  str = str.replace(/\s/g, '')

  for (let i = 0; i < str.length; i++) {
    result += str[i].repeat(i + 1)
  }

  return result
}

// Implementing the Durstenfeld Shuffle
export function scramble(str) {
  str = str.split('')

  for (let i = str.length - 1; i > 0; i--) {
    const swap = Math.floor(Math.random() * (i + 1))
    ;[str[i], str[swap]] = [str[swap], str[i]]
  }

  return str.join('')
}

export function powers(base, limit, callback) {
  let power = 1
  while (power <= limit) {
    callback(power)
    power *= base
  }
}

export function* powersGenerator(base, limit) {
  let power = 1
  while (power <= limit) {
    yield power
    power *= base
  }
}

export function say(str1) {
  const result = []

  function addMoreWords(str2) {
    if (str2 === undefined) {
      return result.join(' ')
    }
    result.push(str2)
    return addMoreWords
  }

  return addMoreWords(str1)
}

export function interleave(arr, ...args) {
  const result = [],
    len = Math.min(arr.length, args.length)

  for (let i = 0; i < len; i++) {
    result.push(arr[i], args[i])
  }

  result.push(...arr.slice(len), ...args.slice(len))

  return result
}

export function makeCryptoFunctions(key, algorithm, iv) {
  const encrypt = decrypted => {
    const cipher = createCipheriv(algorithm, key, iv)
    let encrypted = cipher.update(decrypted, 'utf8', 'hex')
    encrypted += cipher.final('hex')
    return encrypted
  }

  const decrypt = encrypted => {
    const decipher = createDecipheriv(algorithm, key, iv)
    let decrypted = decipher.update(encrypted, 'hex', 'utf8')
    decrypted += decipher.final('utf8')
    return decrypted
  }
  return [encrypt, decrypt]
}

export function topTenScorers(stats) {
  return Object.entries(stats)
    .flatMap(pair => pair[1].map(player => [...player, pair[0]]))
    .filter(player => player[1] >= 15)
    .map(player => {
      return {
        name: player[0],
        ppg: player[2] / player[1],
        team: player[3],
      }
    })
    .sort((player1, player2) => {
      return player2.ppg - player1.ppg
    })
    .slice(0, 10)
}

export async function multiply(x, y) {
  const url = `https://ordinary-hazel-pink.glitch.me/multiply?x=${x}&y=${y}`
  const response = await fetch(url)
  const parsed = await response.json()

  if (!response.ok) {
    throw parsed
  }

  return parsed.result
}
